### ISSUE TITLE
The batchnorm problem about resnet_v2

### ISSUE THREAD

**Issue by HuiyanWen** [tensorflow/models#7722]
## The problem

I tried to run ResNet by using the slim version in **/models/research/slim/nets/resnetv2.py**, but the moving_mean and the moving_variance didn't be saved, even if I did as the official notes:
```
                opt = tf.train.AdamOptimizer(learning_rate=rate)
		update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
		with tf.control_dependencies(update_ops):
			train_op = opt.minimize(loss, global_step=global_step)
		tf.summary.scalar('loss', loss)
		tf.summary.scalar('accuracy', eval_correct)
		for v in tf.global_variables():
			if ('moving_mean' in v.name) or ('moving_variance' in v.name):
				print(v)
				tf.summary.histogram(v.name, v)
```
This phenomenon contributed to a fact that the train accuracy increased rapidly while the test accuracy increased slowly. It seemed like over-fitting, but obviously it was not, because the result was normal when I run resnet by using version v1. This is the log:
```
epoch 0 train set acc:0.0
epoch 0 test accuracy:0.007706925675675676
epoch 1 train set acc:0.39310128348214285
epoch 1 test accuracy:0.2808695972758473
epoch 2 train set acc:0.7352120535714286
epoch 2 test accuracy:0.3768355408980409
epoch 3 train set acc:0.77880859375
epoch 3 test accuracy:0.42969364453739456
epoch 4 train set acc:0.8028041294642857
epoch 4 test accuracy:0.4497834888459889
epoch 5 train set acc:0.8160923549107143
epoch 5 test accuracy:0.45564090876590874
epoch 6 train set acc:0.8241838727678571
epoch 6 test accuracy:0.46629050925925924
epoch 7 train set acc:0.8352399553571429
epoch 7 test accuracy:0.4792214089089089
epoch 8 train set acc:0.844482421875
epoch 8 test accuracy:0.4857497005934506
epoch 9 train set acc:0.8494698660714286
epoch 9 test accuracy:0.47268250393250394
epoch 10 train set acc:0.8593052455357143
epoch 10 test accuracy:0.49918836246961246
epoch 11 train set acc:0.8661411830357143
epoch 11 test accuracy:0.49487322143572143
```
## Environment
The batchnorm function which I used was **slim.batchnorm**, and my python version was 3.6.3, and tensorflow version was 1.7.0

I nearly read all of the methods on the websites, still I couldn't solve it. I'm very appreciated that if you can deal with this problem.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by HuiyanWen** [comment]
> Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
> What is the top-level directory of the model you are using
> Have I written custom code
> OS Platform and Distribution
> TensorFlow installed from
> Bazel version
> CUDA/cuDNN version
> GPU model and memory
> Exact command to reproduce

Thanks for the response @tensorflowbutler. My system is **windows server 2013**, and TensorFlow was installed from **annaconda(CUDA/cuDNN version, GPU model)**,I did write **custom code**, but didn't change the network. Here's the graph definition, and it went well when applied in resnet_v1(without BN version). My problem is why the moving_mean and moving_variance can't be saved even if I've tried as official notes.

### Graph defination
```
images = tf.placeholder(tf.float32, shape=[None, 112,112,1])
labels = tf.placeholder(tf.int32, shape=[None])
lr= tf.placeholder(tf.float32)

#extract features from train samples.
features, _ = resnet_v2_50(images, num_classes=FLAGS.num_classes, is_training=True, reuse=False)

#extract features from validation samples (used for validation during the training).
features_val, _ = resnet_v2_50(images, num_classes=FLAGS.num_classes, is_training=False, reuse=True)

#you can ignore this line, it won't affect the BN.
centers = func.construct_center(features, FLAGS.num_classes)
loss1 = func.dce_loss(features, labels, centers, FLAGS.temp)
loss2 = func.pl_loss(features, labels, centers)
loss = loss1 + FLAGS.weight_pl * loss2

#calculate the accuracy during training.
eval_correct = func.evaluation(features, labels, centers)

#calculate the accuracy during validation.
eval_correct_val = func.evaluation(features_val, labels, centers)

#inference the label for an unknown sample.
predict_index = func.predict(features, centers)
global_step = tf.get_variable("step", [], initializer=tf.constant_initializer(0.0), trainable=False)
rate = tf.train.exponential_decay(FLAGS.learning_rate, global_step, decay_steps=FLAGS.decay_steps, decay_rate=FLAGS.decay_rate,
										  staircase=True)
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS, scope='resnet_v2_50')
updates_op = tf.group(*update_ops)
with tf.control_dependencies([updates_op]):
	train_op = tf.train.AdamOptimizer(learning_rate=rate).minimize(loss, global_step=global_step)
tf.summary.scalar('loss', loss)
tf.summary.scalar('accuracy', eval_correct)
merged_summary_op = tf.summary.merge_all()
```
### Training part
```
while True:
  try:
    # training
    train_images_batch, train_labels_batch = sess.run([train_x, train_y])
    result = sess.run([train_op, loss, eval_correct, global_step, merged_summary_op, predict_index], feed_dict={images:train_images_batch, labels:train_labels_batch})
    pre = result[5] # predict label
    sum_loss += result[1] # loss value
    score = score + result[2]/len(train_labels_batch) # accuracy
    trainIter += 1 
    step = result[3] #global step
    train_writer.add_summary(result[4], step) # write merge_op in tensorboard

    # validate the validation samples evrey FLAGS.eval_steps during training.
    if step % FLAGS.eval_steps == 0:
       print("epoch {0} train set acc:{1}".format(epoch, score/trainIter))
       score = 0.0
       trainIter = 0

       # load the validation set.
       test_x, test_y = get_batch_val(dirpath=FLAGS.val_data_dir, lb='val')
       coord = tf.train.Coordinator()
       while True:
          try:
	     i = 0
	     sum_acc = 0.0
	     while not coord.should_stop():
	         val_images_batch, val_labels_batch = sess.run([test_x, test_y])
		 acc = sess.run(eval_correct_val, feed_dict={images: val_images_batch, labels: val_labels_batch})
		 acc = acc / len(val_labels_batch)
		 sum_acc += acc
		 i+=1
		 except tf.errors.OutOfRangeError:
		 print("epoch {0} accuracy:{1}".format(epoch, sum_acc / (i-1)))
		 epoch+=1
		 break
                          .......
```

I validated the training set during every eval step, the first and second line were both the training set accuracy. This phenomenon illustrates the BN parameters aren't be saved.
### Training log
```
epoch 0 train set acc:0.00390625
epoch 0 accuracy:0.010668649266354961

epoch 1 train set acc:0.39052036830357145
epoch 1 accuracy:0.012217072689778384

epoch 2 train set acc:0.7339564732142857
epoch 2 accuracy:0.01651042854563424

epoch 3 train set acc:0.7799595424107143
epoch 3 accuracy:0.013099679980613526

epoch 4 train set acc:0.79833984375
epoch 4 accuracy:0.008240439806895504
```

**Comment by HuiyanWen** [comment]
@tensorflowbutler Hi, thanks very much, I read this [code](https://github.com/soloice/mnist-bn/blob/master/mnist_bn.py) incidentally. It explains the phenomenon I mentioned above, here it is:
```
# Note: the validation error is erratic in the beginning (Maybe 2~3k steps).
# This does NOT imply the batch normalization is buggy.
# On the contrary, it's BN's dynamics: moving_mean/variance are not estimated that well in the beginning.
```
I've noticed that many people were suffered from this. So if the explanation is correct, maybe you can add some in the official notes about BN. It will help us a lot. Thanks again, and you can close this issue if there's no other problem.
